src/comparison/balanced_pb.rs

```diff
@@ -0,0 +1,214 @@
+    // Actually, there's side effects between the segments in the first loop, so
+    // we might not be able to do it after all.
+    let mut all_history = vec![Vec::new(); segments.len()];
+
+    for attempt in attempts {
+        let attempt_index = attempt.index();
+        let mut history_starting_index = -1;
+
+        for (segment_index, (segment, all_history)) in
+            segments.iter().zip(all_history.iter_mut()).enumerate()
+        {
+            let segment_index = segment_index as i64;
+
+            if let Some(history) = segment.segment_history().get(attempt_index) {
+                if let Some(history) = history[method] {
+                    all_history.push((history_starting_index, history));
+                    history_starting_index = segment_index;
+                }
+            } else {
+                history_starting_index = segment_index;
+            }
+        }
+    }
+
+    // TODO Reuse
+    let mut weighted_lists = Vec::new();
+    let mut overall_starting_index = -1;
+
+    for (current_index, (current_list, segment)) in
+        all_history.into_iter().zip(segments.iter()).enumerate()
+    {
+        let mut null_segment = false;
+        let current_pb_time = segment.personal_best_split_time()[method];
+        let previous_pb_time = if overall_starting_index >= 0 {
+            segments[overall_starting_index as usize].personal_best_split_time()[method]
+        } else {
+            None
+        };
+
+        // TODO More allocations :(
+        let mut final_list = current_list
+            .into_iter()
+            .filter(|&(index, _)| index == overall_starting_index)
+            .map(|(_, time)| time)
+            .collect::<Vec<_>>();
+
+        if !final_list.is_empty() {
+            overall_starting_index = current_index as i64;
+        } else if let Some(diff) = catch! { current_pb_time? - previous_pb_time? } {
+            final_list.push(diff);
+            overall_starting_index = current_index as i64;
+        } else {
+            null_segment = true;
+        }
+
+        if !null_segment {
+            let count = final_list.len();
+            // TODO More allocations FeelsBadMan at this point
+            let mut temp_list = final_list
+                .into_iter()
+                .enumerate()
+                .map(|(i, time)| (get_weight(i, count), time))
+                .collect::<Vec<_>>();
+
+            // TODO Aaarghhh >:(
+            let mut weighted_list = Vec::new();
+
+            if temp_list.len() > 1 {
+                temp_list
+                    .sort_unstable_by_key(|&(_, time)| OrderedFloat(time.total_milliseconds()));
+
+                let total_weight = temp_list.iter().map(|&(weight, _)| weight).sum::<f64>();
+                // TODO That's the smallest time's weight, not the actual
+                // smallest weight. Is that intended?
+                let smallest_weight = temp_list[0].0;
```

Comment:
@ROMaster2 We sort temp_list by the time, but then take the weight of the first element. That's the weight of the smallest time and not the smallest weight in general. Is that intended?